compiler/rustc_infer/src/infer/canonical/mod.rs

```diff
@@ -105,21 +87,26 @@ impl<'tcx> InferCtxt<'tcx> {
         &self,
         span: Span,
         kind: CanonicalVarKind<'tcx>,
+        previous_var_values: &[GenericArg<'tcx>],
         universe_map: impl Fn(ty::UniverseIndex) -> ty::UniverseIndex,
     ) -> GenericArg<'tcx> {
         match kind {
-            CanonicalVarKind::Ty(ty_kind) => {
-                let ty = match ty_kind {
-                    CanonicalTyVarKind::General(ui) => {
-                        self.next_ty_var_in_universe(span, universe_map(ui))
+            CanonicalVarKind::Ty { ui, sub_root } => {
+                let vid = self.next_ty_vid_in_universe(span, universe_map(ui));
+                // Fetch the `sub_root` in case it exists.
+                if let Some(prev) = previous_var_values.get(sub_root.as_usize()) {
```

Comment:
notably this requires that sub roots come *before* any vars that are related to them, and if we muck that up then we implicitly wind up "dropping" that information. this presumably cant lead to soundness bugs because we dont use sub_relation for soundness critical stuff :> 

worst case here is just we wind up with less sub relations than we'd like and so opaque types work worse :tm: are we able to like. explode. a bit more loudly, if this goes wrong? is the non-existent-`else` on this `if let` actually reachable in some cases?